### DISCUSSION TITLE
[Idea] Multiple concurrent PHP versions

### DISCUSSION THREAD

**Discussion by atymic** [laravel/valet#1129] [Ideas]
I have a bunch of different client projects running different versions of PHP. I love valet, it's so much more light weight that docker but the only thing that bugs me is constantly switching php versions. 

I was wondering if you would consider a PR that allows you pin a site to specific  FPM version (having multiple running concurrently)?

I will take a crack on my fork :)

**Comment by jjpmann** [comment]
I normally just secure the site with 'valet secure' so I can modify the nginx.conf file. You can use this as an example: https://github.com/laravel/valet/pull/941

I once overloaded `valet_support_wildcard_dns` function to allow custom domain naming for PHP versions so I could map nginx to those. If interested I can dig up those config changes needed.

Example:
site.test - default
site.php56.test - run php5.6
site.php70.test - run php7.0

**Comment by atymic** [comment]
Ahh, that's not a bad idea.

I was doing something similar, where you do `valet site-use php@7.x` and it just create/updates the nginx config to use the correct FPM version (plus making sure the service is running)

**Comment by vesper8** [comment]
@atymic I love your idea and am also having to switch php versions often and it's very annoying.. specially since there seems to be a bug when switching back to php74 and everytime it has to be reinstalled and it complains that it isn't already installed even though it clearly is

any progress on your fork? @mattstauffer Would love to see this `valet site-use` feature make it into the core

**Comment by yasirmturk** [comment]
Can we modify 
`/usr/local/etc/nginx/valet/valet.conf` or `~/.config/valet/Nginx/<site>.text` to point all request to php-fpm running inside a docket container while keeping valet/apache on local machine.. this is more relevant when using `valet secure`
for example `fastcgi_pass 127.0.0.1:9000;` ???

**Comment by BurningDog** [comment]
https://dev.to/dyriavin/serving-sites-with-different-php-version-using-laravel-valet-49jl is a pretty helpful guide. I've made a slight addition to have a default php version - 8.1 - and another version available by manually editing the site's nginx config file in `~/.config/valet/Nginx`.

Install both versions of php:

```sh
brew install php@7.4
brew unlink php@7.4
brew install php
```

`php -v` should show `php 8`. I haven't bothered setting up aliases for the different php versions when used from the command line, I just do:

```sh
brew unlink php
brew link php@7.4
```

This means php 7.4 is now the active php (in my current - and new - terminal tabs). Other open terminal tabs will keep whatever version of php was active there.

Edit `/opt/homebrew/etc/php/7.4/php-fpm.d/valet-fpm.conf`: change `listen = /Users/**user-name**/.config/valet/valet.sock` to `listen = /Users/**user-name**/.config/valet/valet74.sock`

Make sure the `php@7.4` service is running:

```
brew services
brew services start php@7.4
```

Restart valet: `valet restart`

All served sites will now be php 8. To set a specific site to php 7.4, `cd ~/.config/valet/Nginx` and edit the site. Find the following line:

```
fastcgi_pass "unix:/Users/**user-name**/.config/valet/valet.sock";
```
and change it to
```
fastcgi_pass "unix:/Users/**user-name**/.config/valet/valet74.sock";
```

Then restart valet: `valet restart`.
